src/librustc_trans/back/msvc/mod.rs

```diff
@@ -46,283 +46,179 @@ pub fn link_exe_cmd(sess: &Session) -> Command {
-            })
-        }).or_else(|| {
-            env::var_os("VS120COMNTOOLS")
-        }).or_else(|| {
-            env::var_os("VS100COMNTOOLS")
-        }).or_else(|| {
-            env::var_os("VS90COMNTOOLS")
-        }).or_else(|| {
-            env::var_os("VS80COMNTOOLS")
-        }).map(PathBuf::from).and_then(|mut dir| {
-            if dir.ends_with("Common7/IDE") || dir.ends_with("Common7/Tools") {
-                dir.pop();
-                dir.pop();
-                Some(dir)
-            } else {
-                None
-            }
+    // Sorting the directories asciibetically isn't ideal, but it is what
+    // vcvars does so that's good enough for us
+    fn get_ucrt_dir() -> Option<PathBuf> {
+        LOCAL_MACHINE.open(r"SOFTWARE\Microsoft\Windows Kits\Installed Roots".as_ref()).ok()
+        .and_then(|key| {
+            key.query_str("KitsRoot10").ok()
+        }).and_then(|root| {
+            fs::read_dir(Path::new(&root).join("Lib")).ok()
+        }).and_then(|readdir| {
+            let mut dirs: Vec<_> = readdir.filter_map(|dir| dir.ok())
+                .map(|dir| dir.path()).collect();
+            dirs.sort();
+            dirs.pop()
         })
     }
 
-    // Given a registry key, look at all the sub keys and find the one which has
-    // the maximal numeric value.
-    //
-    // Returns the name of the maximal key as well as the opened maximal key.
-    fn max_version(key: &RegistryKey) -> Option<(OsString, RegistryKey)> {
-        let mut max_vers = 0;
-        let mut max_key = None;
-        for subkey in key.iter().filter_map(|k| k.ok()) {
-            let val = subkey.to_str().and_then(|s| {
-                s.trim_left_matches("v").replace(".", "").parse().ok()
-            });
-            let val = match val {
-                Some(s) => s,
-                None => continue,
-            };
-            if val > max_vers {
-                if let Ok(k) = key.open(&subkey) {
-                    max_vers = val;
-                    max_key = Some((subkey, k));
-                }
-            }
-        }
-        return max_key
+    // vcvars finds the correct version of the Windows 10 SDK by looking
+    // for the include um/Windows.h because sometimes a given version will
+    // only have UCRT bits without the rest of the SDK.
+    // Since we only care about libraries and not includes, we just look
+    // for the folder `um` in the lib section.
+    fn get_sdk10_dir() -> Option<PathBuf> {
+        LOCAL_MACHINE.open(r"SOFTWARE\Microsoft\Microsoft SDKs\Windows\v10.0".as_ref()).ok()
+        .and_then(|key| {
+            key.query_str("InstallationFolder").ok()
+        }).and_then(|root| {
+            fs::read_dir(Path::new(&root).join("Lib")).ok()
+        }).and_then(|readdir| {
+            let mut dirs: Vec<_> = readdir.filter_map(|dir| dir.ok())
+                .map(|dir| dir.path()).collect();
+            dirs.sort();
+            dirs.reverse();
```

Comment:
Instead of `reverse` here you should also be able to do `dirs.into_iter().rev().filter(...)` below